pkg/api/write.go

```diff
@@ -148,6 +158,96 @@ func validateWriteHeaders(w http.ResponseWriter, r *http.Request) bool {
 }
 
 func buildWriteError(w http.ResponseWriter, err string) {
-	log.Error("msg", "Write header validation error", "err", err)
+	buildError(w, err, "Write header validation error")
+}
+
+func buildError(w http.ResponseWriter, err string, msg string) {
+	log.Error("msg", msg, "err", err)
 	http.Error(w, err, http.StatusBadRequest)
 }
+
+func buildImportError(w http.ResponseWriter, err string) {
+	buildError(w, err, "Import header validation error")
+}
+
+type importPayload struct {
+	Metric     map[string]string `json:"metric"`
+	Timestamps []int64           `json:"timestamps"`
+	Values     []float64         `json:"values"`
+}
+
+func Import(writer pgmodel.DBInserter, elector *util.Elector, metrics *Metrics) http.Handler {
+	return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+		if !validateImportHeaders(w, r) {
+			return
+		}
+
+		if !checkLeader(elector, metrics) {
+			return
+		}
+
+		var i importPayload
+		dec := json.NewDecoder(r.Body)
+		req := pgmodel.NewWriteRequest()
+
+		for {
+			if err := dec.Decode(&i); err == io.EOF {
+				break
+			} else if err != nil {
+				log.Error("msg", "Import error", "err", err.Error())
+				http.Error(w, err.Error(), http.StatusInternalServerError)
+				return
+			}
+
+			if len(i.Timestamps) != len(i.Values) {
+				log.Error("msg", "Import error", "err", "Mismatch number of timestamps and values in payload")
+				http.Error(w, "Mismatch number of timestamps and values in payload", http.StatusInternalServerError)
+				return
+			}
```

Comment:
This feels like a bit of an awkward constraint to impose; is this better than having the input be an array of `(time, value)` pairs?